app/queues/dns/create-record-worker.server.ts

```diff
@@ -0,0 +1,142 @@
+import { redis } from '~/lib/redis.server';
+import { prisma } from '~/db.server';
+import logger from '~/lib/logger.server';
+import { setMonthsFromNow } from '~/lib/domains.server';
+import { createRecord, getChangeStatus } from '~/lib/dns.server';
+import { RecordStatus } from '@prisma/client';
+
+import type { DomainRecord } from '~/lib/domains.server';
+
+export const addDbRecordQueueName = 'add-db-record';
+export const createDnsRecordQueueName = 'create-dns-record';
+export const checkDnsStatusQueueName = 'check-dns-status';
+export const syncDbStatusQueueName = 'sync-db-status';
+
+export const addDbRecordWorker = new Worker<
+  Pick<DomainRecord, 'username' | 'type' | 'name' | 'value'>
+>(
+  addDbRecordQueueName,
+  async (job) => {
+    logger.info('Creating a record in DB');
+    const { username, type, name, value } = job.data;
+
+    if (!username) {
+      throw new Error('Username is not provided');
+    }
+
+    const existingRecords = await prisma.record.count({
+      where: {
+        type,
+        name,
+        value,
+      },
+    });
+
+    if (existingRecords > 0) {
+      throw new Error('Record already exists');
+    }
+
+    try {
+      const result = await prisma.record.create({
+        data: {
+          username,
+          type,
+          name,
+          value,
+          expiresAt: setMonthsFromNow(6),
+          status: RecordStatus.pending,
+        },
+      });
+      return result.id;
+    } catch (error) {
+      logger.warn('Could not create a record in DB', { error });
+    }
+  },
+  { connection: redis }
+);
+
+export const createDnsRecordWorker = new Worker<
+  Pick<DomainRecord, 'username' | 'type' | 'name' | 'value'>
+>(
+  createDnsRecordQueueName,
+  async (job) => {
+    logger.info('Creating a record in Route 53');
+    const { username, type, name, value } = job.data;
+
+    try {
+      return await createRecord(username, type, name.toLowerCase(), value);
+    } catch (error) {
+      logger.warn('Could not create a record in Route53', { error });
+    }
+  },
+  { connection: redis }
+);
+
+export const checkDnsStatusWorker = new Worker(
+  checkDnsStatusQueueName,
+  async (job) => {
+    logger.info('Checking Route53 readiness');
+
+    const values: any = await job.getChildrenValues();
+    const key = Object.keys(values)[0];
+    const changeId = values[key];
+    let status: string;
+
+    do {
+      status = await getChangeStatus(changeId);
+      await delay(1000);
+    } while (status === 'PENDING');
+    return status;
+  },
+  { connection: redis }
+);
+
+export const syncDbStatusWorker = new Worker(
+  syncDbStatusQueueName,
+  async (job) => {
+    logger.info('Sync DB status with DNS');
+    const values = await job.getChildrenValues();
+    const keys = Object.keys(values);
+
+    let recordId, status;
+    if (keys.length !== 2) {
+      throw new Error('Invalid children values from previous workers');
+    } else {
+      if (keys[0].includes(addDbRecordQueueName)) {
```

Comment:
Not sure if I understand this part. What does this do?

Reply:
In my testing, values from `job.getChildrenValues()` always returns the values in order. So this is not necessary in my experience so far. But I decided to assume that it can be a different order just to be safe. So the first value in the object can be either one of `dbId` or `status`.

Reply:
Thanks for the explanation

To be honest this looks a bit flaky, and hard(er) to maintain if the job dependency chain chnages. What I'd do is:
```js
    const childJobResults = await job.getChildrenValues();
    const addRecordJobKey = Object.keys(childJobResults).find(item => item.includes(addDbRecordQueueName))
    const recordId = childJobResults[addRecordJobKey]
```

This way no matter what the order was, how many steps you had before, you always get the correct result

This is just a suggestion, so if you don't agree, feel free to ignore it

Reply:
It's true that my code is less maintainable. Thanks for the suggestion.

Reply:
Happy if it helped :)